src/main/java/appeng/me/storage/MEMonitorIFluidHandler.java

```diff
@@ -0,0 +1,327 @@
+		this.memory = new ConcurrentSkipListMap<>();
+	}
+
+	@Override
+	public void addListener( final IMEMonitorHandlerReceiver<IAEFluidStack> l, final Object verificationToken )
+	{
+		this.listeners.put( l, verificationToken );
+	}
+
+	@Override
+	public void removeListener( final IMEMonitorHandlerReceiver<IAEFluidStack> l )
+	{
+		this.listeners.remove( l );
+	}
+
+	@Override
+	public IAEFluidStack injectItems( final IAEFluidStack input, final Actionable type, final IActionSource src )
+	{
+		final int filled = this.handler.fill( input.getFluidStack(), type == type.MODULATE );
+
+		if( filled == 0 )
+		{
+			return input.copy();
+		}
+
+		if( type == Actionable.MODULATE )
+		{
+			this.onTick();
+		}
+
+		if( filled == input.getStackSize() )
+		{
+			return null;
+		}
+
+		final IAEFluidStack o = input.copy();
+		o.setStackSize( input.getStackSize() - filled );
+		return o;
+	}
+
+	@Override
+	public IAEFluidStack extractItems( final IAEFluidStack request, final Actionable type, final IActionSource src )
+	{
+		final FluidStack removed = this.handler.drain( request.getFluidStack(), type == Actionable.MODULATE );
+
+		if( removed == null || removed.amount == 0 )
+		{
+			return null;
+		}
+
+		if( type == Actionable.MODULATE )
+		{
+			this.onTick();
+		}
+
+		final IAEFluidStack o = request.copy();
+		o.setStackSize( removed.amount );
+		return o;
+	}
+
+	@Override
+	public IStorageChannel getChannel()
+	{
+		return AEApi.instance().storage().getStorageChannel( IFluidStorageChannel.class );
+	}
+
+	@Override
+	public TickRateModulation onTick()
+	{
+		final List<IAEFluidStack> changes = new ArrayList<>();
+
+		this.list.resetStatus();
+		int high = 0;
+		boolean changed = false;
+
+		final IFluidTankProperties[] props = this.handler.getTankProperties();
+		for( int slot = 0; slot < this.handler.getTankProperties().length; ++slot )
+		{
+			final CachedFluidStack old = this.memory.get( slot );
+			high = Math.max( high, slot );
+
+			final FluidStack newIS = !props[slot].canDrain() && this.getMode() == StorageFilter.EXTRACTABLE_ONLY ? null : props[slot].getContents();
+			final FluidStack oldIS = old == null ? null : old.fluidStack;
+
+			if( isDifferent( newIS, oldIS ) )
+			{
+				final CachedFluidStack cis = new CachedFluidStack( newIS );
+				this.memory.put( slot, cis );
+
+				if( old != null && old.aeStack != null )
+				{
+					old.aeStack.setStackSize( -old.aeStack.getStackSize() );
+					changes.add( old.aeStack );
+				}
+
+				if( cis.aeStack != null )
+				{
+					changes.add( cis.aeStack );
+					this.list.add( cis.aeStack );
+				}
+
+				changed = true;
+			}
+			else
+			{
+				final int newSize = newIS == null ? 0 : newIS.amount;
+				final int diff = newSize - ( oldIS == null ? 0 : oldIS.amount );
+
+				final IAEFluidStack stack = ( old == null || old.aeStack == null ? AEApi.instance()
+						.storage()
+						.getStorageChannel( IFluidStorageChannel.class )
+						.createStack( newIS ) : old.aeStack.copy() );
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Parameter 1 to this call is marked "javax.annotation.Nonnull" but null could be passed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.orod.org/coding_rules#rule_key=squid%3AS2637)